### PR TITLE
A ribbon rule integrated with nacos cluster and respect nacos server'…

### PR DESCRIPTION
**Pull request by eacdy**
A ribbon rule integrated with nacos cluster and respect nacos server's weight configuration.

### CODE REVIEW

File: spring-cloud-alibaba-nacos-discovery/src/main/java/org/springframework/cloud/alibaba/nacos/ribbon/ExtendBalancer.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import com.alibaba.nacos.client.naming.core.Balancer;
+
+import java.util.List;
+
+/**
+ * @author itmuch.com
+ */
+public class ExtendBalancer extends Balancer {
+    /**
+     * 根据权重，随机选择实例
```
Comment (fangjian0423) [inline spring-cloud-alibaba-nacos-discovery/src/main/java/org/springframework/cloud/alibaba/nacos/ribbon/ExtendBalancer.java]: we should use english comments

File: spring-cloud-alibaba-nacos-discovery/src/main/java/org/springframework/cloud/alibaba/nacos/ribbon/NacosRule.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+            List<Instance> instancesToChoose = instances;
+
+            if (StringUtils.isNotBlank(clusterName)) {
+                List<Instance> sameClusterInstances = instances.stream()
+                        .filter(instance -> Objects.equals(clusterName, instance.getClusterName()))
+                        .collect(Collectors.toList());
+                if (!CollectionUtils.isEmpty(sameClusterInstances)) {
+                    instancesToChoose = sameClusterInstances;
+                } else {
+                    LOGGER.warn("发生跨集群的调用，name = {}, clusterName = {}, instance = {}", name, clusterName, instances);
```
Comment (fangjian0423) [inline spring-cloud-alibaba-nacos-discovery/src/main/java/org/springframework/cloud/alibaba/nacos/ribbon/NacosRule.java]: we should use english comments

File: spring-cloud-alibaba-nacos-discovery/src/main/java/org/springframework/cloud/alibaba/nacos/ribbon/NacosRule.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+                }
+            }
+
+            Instance instance = ExtendBalancer.getHostByRandomWeight2(instancesToChoose);
```
Comment (fangjian0423) [inline spring-cloud-alibaba-nacos-discovery/src/main/java/org/springframework/cloud/alibaba/nacos/ribbon/NacosRule.java]: NPE check

File: spring-cloud-alibaba-nacos-discovery/src/main/java/org/springframework/cloud/alibaba/nacos/ribbon/NacosRule.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+            return new NacosServer(instance);
```
Comment (fangjian0423) [inline spring-cloud-alibaba-nacos-discovery/src/main/java/org/springframework/cloud/alibaba/nacos/ribbon/NacosRule.java]: if instance is null, will throw NPE

**Comment by fangjian0423** [comment]
circleci build error.

```java
log.warn("{}服务当前无任何实例", name);
```

it seems the log is not defined.

**Comment by eacdy** [comment]
> circleci build error.
> 
> ```java
> log.warn("{}服务当前无任何实例", name);
> ```
> 
> it seems the log is not defined.

My fault. I use lombok in my local machine...

**Pull request merged**